src/Shared/HttpSys/RequestProcessing/RequestHeaders.cs

```diff
@@ -82,7 +79,17 @@ public bool ContainsKey(string key)
 
     public ICollection<string> Keys
     {
-        get { return PropertiesKeys().Concat(Extra.Keys).ToArray(); }
+        get
+        {
+            var destination = new string[Count];
+            Span<byte> observedKnownHeaders = stackalloc byte[HeaderKeys.Length];
+            int observedKnownHeadersCount = _requestMemoryBlob.GetKeys(HeaderKeys, observedKnownHeaders, destination);
```

Comment:
This doesn't account for the headers collection being modified like in https://github.com/dotnet/aspnetcore/pull/44860. A removed header will still be listed in keys, and an added header won't be. This might need to build on the logic being added for the other PR.

Count has the same issue.

Reply:
Good point. I'll wait for that PR

Reply:
Only known headers are removable?

Reply:
This should be resolved now.